### PR TITLE
refactor(optimizer): unify watermark derivation and monotonicity derivation

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Previously we have

1. Watermark derivation to derive output watermark columns for each plan node;
2. `Nondecreasing` variant of `WatermarkDerivation` to support watermark on `proctime` column;
3. Another attempt #13984 to implement another property of columns of plan node called "monotonicity".

When I worked on the continuation of the "monotonicity" property, I realized that we can unify the watermark and monotonicity derivation process, by

1. Analyze the `WatermarkDerivation` in the same recursive manner as we do for `WatermarkDerivation`;
2. Derive watermark based on the `WatermarkDerivation`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/property/monotonicity.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+/// Represents the monotonicity of a column. This enum aims to unify the "non-decreasing analysis" and watermark derivation.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EnumAsInner)]
+pub enum Monotonicity {
+    /// The column is constant.
+    Constant,
+    /// The column is strictly non-decreasing. Basically this is similar to `WATERMARK FOR col AS col`, but the `Watermark`
+    /// messages are not supposed to be produced by `WatermarkFilter`.
+    Increasing,
+    /// The column is strictly non-increasing.
+    Decreasing,
```
Comment (st1page) [inline src/frontend/src/optimizer/property/monotonicity.rs:30]: Placing these two words together seems odd. As far as I understand, there are only "non-increasing" or "strictly increasing". Perhaps defining the enum directly as `NonIncreasing` and `NonDecreasing` would be clearer

File: src/frontend/src/optimizer/property/monotonicity.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    /// The column is GENERALLY non-decreasing, but not STRICTLY. Watermarks should be produced by `WatermarkFilter` and
+    /// forwarded to downstream with necessary transformation in other operators like `Project(Set)`.
+    /// This is not used currently.
+    _IncreasingByWatermark,
```
Comment (st1page) [inline src/frontend/src/optimizer/property/monotonicity.rs:34]:
```suggestion
    _IncreasingByWatermark(usize),
```
I have considered this situation but cannot provide an example. LGTM to keep it

**Review by st1page — COMMENTED** [review]
general LGTM

**Review by chenzl25 — APPROVED** [review]
LGTM

**Review by st1page — APPROVED** [review]

**Pull request merged**